### PR TITLE
pilot: unify pilot cert provider for agent and discovery

### PR DESCRIPTION
**Pull request by kailun-qin**
Signed-off-by: Kailun Qin <kailun.qin@intel.com>

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[x] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pkg/istio-agent/agent.go [hunk 1]
```diff
@@ -435,12 +435,14 @@ func (a *Agent) FindRootCAForXDS() (string, error) {
...
 		rootCAPath = a.cfg.XDSRootCerts
-	} else if fileExists("./etc/certs/root-cert.pem") {
+	} else if fileExists(security.DefaultRootCertFilePath) {
 		// Old style - mounted cert. This is used for XDS auth only,
 		// not connecting to CA_ADDR because this mode uses external
 		// agent (Secret refresh, etc)
-		return "./etc/certs/root-cert.pem", nil
-	} else if a.secOpts.PilotCertProvider == "kubernetes" {
+		return security.DefaultRootCertFilePath, nil
+	} else if a.secOpts.PilotCertProvider == constants.CertProviderNone {
```
Comment (howardjohn) [inline pkg/istio-agent/agent.go:443]: should this be last? Seems like if you use file mounted certs or prov certs its fine to have this be "none"

File: pilot/pkg/bootstrap/certcontroller.go [hunk 2]
```diff
@@ -166,8 +167,13 @@ func (s *Server) initDNSCerts(hostname, customHost, namespace string) error {
 			}
 		}
 	} else {
-		log.Infof("User specified cert provider: %v", features.PilotCertProvider.Get())
-		return nil
+		customCACertPath := security.DefaultRootCertFilePath
```
Comment (lei-tang) [inline pilot/pkg/bootstrap/certcontroller.go:170]: As PilotCertProvider can be "kubernetes", "istiod", "custom" and "none", should there be a "else if features.PilotCertProvider.Get() == constants.CertProviderNone" section?

File: pkg/istio-agent/agent.go (hunk 1 above)
Comment (kailun-qin) [reply pkg/istio-agent/agent.go]: Thanks! Make sense to me. Changed to the last of the specified in terms of ordering.

File: pkg/istio-agent/agent.go [hunk 3]
```diff
@@ -474,26 +476,39 @@ func fileExists(path string) bool {
...
+func (a *Agent) FindRootCAForCA() (string, error) {
+	var rootCAPath string
+
 	if a.cfg.CARootCerts == security.SystemRootCerts {
-		return ""
+		return "", nil
 	} else if a.cfg.CARootCerts != "" {
-		return a.cfg.CARootCerts
-	} else if a.secOpts.PilotCertProvider == "kubernetes" {
+		rootCAPath = a.cfg.XDSRootCerts
```
Comment (lei-tang) [inline pkg/istio-agent/agent.go:485]: When a.cfg.CARootCerts != "", why does the PR change the root CA path from a.cfg.CARootCerts to a.cfg.XDSRootCerts?

Comment (kailun-qin) [reply pkg/istio-agent/agent.go:485]: Good catch! Should be CARootCerts, thanks!

File: pilot/pkg/bootstrap/certcontroller.go (hunk 2 above)
Comment (kailun-qin) [reply pilot/pkg/bootstrap/certcontroller.go:170]: The none case should've been ruled out by https://github.com/istio/istio/blob/master/pilot/pkg/bootstrap/server.go#L903-L905 and https://github.com/istio/istio/blob/master/pilot/pkg/bootstrap/server.go#L931-L933.

File: pkg/istio-agent/agent.go [hunk 4]
```diff
@@ -474,26 +476,39 @@ func fileExists(path string) bool {
...
-		return a.secOpts.ProvCert + "/root-cert.pem"
+		// For VMs, the root cert file used to auth may be populated afterwards.
+		// Thus, return directly here and skip checking for existence.
+		return a.secOpts.ProvCert + "/root-cert.pem", nil
+	} else if a.secOpts.PilotCertProvider == constants.CertProviderNone {
+		return "", fmt.Errorf("root CA file for CA required but configured provider as none")
 	} else {
 		// This is the default - a mounted config map on K8S
-		return path.Join(CitadelCACertPath, constants.CACertNamespaceConfigMapDataName)
+		rootCAPath = path.Join(CitadelCACertPath, constants.CACertNamespaceConfigMapDataName)
```
Comment (lei-tang) [inline pkg/istio-agent/agent.go:502]: The original code does not check the existence of the root cert file for the mounted config map (the original code just returns the path). The PR checks the existence of the root cert file for the mounted config map. Quoted from the comment "// For VMs, the root cert file used to auth may be populated afterwards", similar to the root cert file for VM, is it possible that the root cert file for the mounted config map may be populated afterwards?

Comment (kailun-qin) [reply pkg/istio-agent/agent.go:502]: IMO, I think this should be available as part of the startup. Any comment? @howardjohn @ramaraochavali

Comment (ramaraochavali) [reply pkg/istio-agent/agent.go:502]: I am not aware of how this config map works.

**Review by howardjohn — APPROVED** [review]

**Pull request merged**